### PR TITLE
fix(code-mappings): Fixes incorrect logic when finding stack trace and source path roots 

### PR DESCRIPTION
**Pull request by MichaelSun48**
This PR fixes an issue related to the logic for deriving the stack trace root and source code root for a code-mapping being different when code mappings were derived automatically versus manually. 

Previously, when a user clicked on the "Set up code mapping" button on their stack trace and inputted a valid Github repo to link the stack trace to, AND the roots were shared between stack trace and source path, then the stack trace root would be converted to a "./", and the source code root would be converted to "/". This is erroneous, and inconsistent with the behavior when code mappings are derived manually. This behavior was introduced 3 weeks ago in #65474. 

This PR fixes the logic and tests for manual code mappings to reflect that of automatic code mappings. 

This PR also introduces the following small refactors:

1.  Moves the `find_roots` function to the `integrations/utils/code_mappings.py` file since this is a utility function used in a Github integration. Yes, this means that there are two different functions with slightly different behaviors in code_mappings.py used to derive the source path and stack trace roots. This logic will be unified in an upcoming PR to release Go automatic code mappings, but this is a more complicated fix and is out of the scope of a quick fix PR like this. 
~~2. Changes the `projects/<org_slug>/<project_slug>/repo-path-parsing/` endpoint from a POST to a GET. This endpoint only returns data it retrieves and does not add or change any data, thus it should be a GET. (AFAICT, this endpoint is not public, so no need for docs changes)~~
~~3. Changes all usages of this^ endpoint on the frontend to be a GET request. ~~

This endpoint has to be a post since it takes in a URL and stack trace in the request body, which shouldn't be turned into a query param

### CODE REVIEW

File: src/sentry/api/endpoints/project_repo_path_parsing.py [hunk 1]
```diff
@@ -11,41 +11,10 @@
...
-    overlap_to_check = stack_path.split(stack_path_delim)
-    stack_root = []
-    while overlap_to_check:
-        if source_path.endswith(overlap := SLASH.join(overlap_to_check)):
-            source_root = source_path.rpartition(overlap)[0]
-            stack_root = stack_path_delim.join(stack_root)
-            if not source_root:  # replace empty source root with "slash"
-                source_root = SLASH
-            if not stack_root:  # replace empty stack root with "dot slash"
-                stack_root = f".{stack_path_delim}"
```
Comment (MichaelSun48) [inline src/sentry/api/endpoints/project_repo_path_parsing.py:36]: These two lines were removed to reflect the fact that source_root and stack_root should be empty when roots match

**Comment by MichaelSun48** [comment]
> 🚨 **Warning:** This pull request contains Frontend and Backend changes!
> 
> It's discouraged to make changes to Sentry's Frontend and Backend in a single pull request. The Frontend and Backend are **not** atomically deployed. If the changes are interdependent of each other, they **must** be separated into two pull requests and be made forward or backwards compatible, such that the Backend or Frontend can be safely deployed independently.
> 
> Have questions? Please ask in the [`#discuss-dev-infra` channel](https://app.slack.com/client/T024ZCV9U/CTJL7358X).

How should I go about addressing this issue? Since frontend deploys faster, there will be a short period between when frontend deploys and backend is deploying where the request will not work. 

This endpoint is only called when the user clicks save on this modal: 
image

File: src/sentry/api/endpoints/project_repo_path_parsing.py [hunk 2]
```diff
@@ -122,7 +91,7 @@ class ProjectRepoPathParsingEndpointLoosePermission(ProjectPermission):
 @region_silo_endpoint
 class ProjectRepoPathParsingEndpoint(ProjectEndpoint):
     publish_status = {
-        "POST": ApiPublishStatus.UNKNOWN,
+        "GET": ApiPublishStatus.UNKNOWN,
```
Comment (armenzg) [inline src/sentry/api/endpoints/project_repo_path_parsing.py:94]: Can you please change it to private?

Why the change from `POST` to `GET`? I don't see a `get` method in the endpoint.

I don't fully grasp the meaning of `publish_status` is, thus, you may be right with this change.

File: src/sentry/api/endpoints/project_repo_path_parsing.py (hunk 1 above)
Comment (armenzg) [reply src/sentry/api/endpoints/project_repo_path_parsing.py:36]: What I tried mentioning in the call (I possibly did not say it or I explained incorrectly) is to have first a PR for the fix so it's easy to see what gets changed and a follow-up PR for any refactoring. This makes it easier for the reviewer to see what changes.

Don't worry about it. I will review it as is

File: src/sentry/integrations/utils/code_mapping.py [hunk 3]
```diff
@@ -535,3 +537,29 @@ def _get_code_mapping_source_path(src_file: str, frame_filename: FrameFilename)
     if source_code_root:
         assert source_code_root.endswith("/")
     return source_code_root
+
+
+def find_roots(stack_path, source_path) -> tuple[str, str]:
```
Comment (armenzg) [inline src/sentry/integrations/utils/code_mapping.py:542]:
```suggestion
def find_roots(stack_path: str, source_path: str) -> tuple[str, str]:
```

When the full signature is typed, then mypy does a better job at typing rather than assuming `Any`.

File: src/sentry/integrations/utils/code_mapping.py [hunk 4]
```diff
@@ -535,3 +537,29 @@ def _get_code_mapping_source_path(src_file: str, frame_filename: FrameFilename)
...
+                stack_root = f"{stack_root}{stack_path_delim}"
+
+            return (stack_root, source_root)
+
+        # increase stack root specificity, decrease overlap specifity
+        stack_root_items.append(overlap_to_check.pop(0))
+
+    # validate_source_url should have ensured the file names match
+    # so if we get here something went wrong and there is a bug
+    raise Exception("Could not find common root from paths")
```
Comment (armenzg) [inline src/sentry/integrations/utils/code_mapping.py:565]: Do we know the behaviour in the UI when we raise this exception?

File: tests/sentry/api/endpoints/test_project_repo_path_parsing.py [hunk 5]
```diff
@@ -197,7 +197,7 @@ def test_basic(self):
             "integrationId": self.integration.id,
             "repositoryId": self.repo.id,
             "provider": "github",
-            "stackRoot": "./",
+            "stackRoot": "",
```
Comment (armenzg) [inline tests/sentry/api/endpoints/test_project_repo_path_parsing.py:200]: Yay!

File: src/sentry/api/endpoints/project_repo_path_parsing.py (hunk 1 above)
Comment (armenzg) [reply src/sentry/api/endpoints/project_repo_path_parsing.py:36]: For anyone coming along, this changes the addition of these lines:
https://github.com/getsentry/sentry/pull/65474/files#diff-08924ad13c52b755328b00053d214294721782bac7af4600e3f3ad4599b3180bR33-R38

File: src/sentry/integrations/utils/code_mapping.py (hunk 4 above)
Comment (MichaelSun48) [reply src/sentry/integrations/utils/code_mapping.py:565]: image

File: src/sentry/api/endpoints/project_repo_path_parsing.py (hunk 2 above)
Comment (MichaelSun48) [reply src/sentry/api/endpoints/project_repo_path_parsing.py:94]: Typically endpoints labeled as "GET" only retrieve and return data from somewhere on our backend, as opposed to "POST" endpoints that actually adds or modifies data in our backend. Since this method doesn't change or add data in our backend, it's much better labeled as a GET endpoint.

For a private endpoint like this, the difference doesn't functionally matter, but I personally think should still be changed for understandability.  

You're right that there is no `get` method in the endpoint though - this is a mistake that I'll fix right now.

File: src/sentry/api/endpoints/project_repo_path_parsing.py (hunk 1 above)
Comment (MichaelSun48) [reply src/sentry/api/endpoints/project_repo_path_parsing.py:36]: Ah I see what you mean now, I realize might've jumped the gun on a couple of these changes and they could've waited for a future PR... sorry about that!

File: src/sentry/api/endpoints/project_repo_path_parsing.py (hunk 2 above)
Comment (MichaelSun48) [reply src/sentry/api/endpoints/project_repo_path_parsing.py:94]: (I suspect the absence of a `get` method is what is causing that error)

**Review by armenzg — APPROVED** [review]
It's looking good. Once you get the tests passing you can get it merged. Don't merge it late today.
